### PR TITLE
arithmetic overflow

### PR DESCRIPTION
**Pull request by siddontang**
+ check integer divide overflow 
+ the unsigned/signed return type for MOD depends on first operand.

### CODE REVIEW

File: util/types/overflow.go [hunk 1]
```diff
@@ -151,3 +151,39 @@ func MulInteger(a uint64, b int64) (uint64, error) {
 
 	return MulUint64(a, uint64(b))
 }
+
+// DivInt64 divide int64 a with b, returns int64 if no overflow error.
```
Comment (qiuyesuifeng) [inline util/types/overflow.go]: divides

File: util/types/overflow.go [hunk 2]
```diff
@@ -151,3 +151,39 @@ func MulInteger(a uint64, b int64) (uint64, error) {
...
+			return 0, errors.Trace(ErrArithOverflow)
+		}
+
+		return 0, nil
+	}
+
+	return a / uint64(b), nil
+}
+
+// DivIntWithUint divide int64 a with uint64 b, returns uint64 if no overflow error.
```
Comment (qiuyesuifeng) [inline util/types/overflow.go]: divides

File: util/types/overflow_test.go [hunk 3]
```diff
@@ -263,3 +263,73 @@ func (s *testOverflowSuite) TestMul(c *C) {
...
+		ret      uint64
+		overflow bool
+	}{
+		{math.MinInt64, math.MaxInt64, 0, true},
+		{0, 1, 0, false},
+		{-1, math.MaxInt64, 0, false},
+	}
+
+	for _, t := range tblInt2 {
+		c.Log(t.lsh, t.rsh)
```
Comment (qiuyesuifeng) [inline util/types/overflow_test.go]: this line can be removed?

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**